src/odemis/acq/align/fastem.py

```diff
@@ -18,22 +20,288 @@
+            self.asm_config["multibeam"]["scanOffset"] = s_offset
+            self.asm_config["multibeam"]["scanAmplitude"] = s_amplitude
+            configure_asm(self._multibeam, self._descanner, self._detector, self._dataflow, self.asm_config,
+                          upload=False)
+
+        elif calibration == Calibrations.DESCAN_GAIN_STATIC:
+            descan_gain.run_descan_gain_static(self._scanner, self._multibeam, self._descanner,
+                                               self._detector, self._dataflow, self._ccd)
+            configure_asm(self._multibeam, self._descanner, self._detector, self._dataflow, self.asm_config,
+                          upload=False)
+
+        elif calibration == Calibrations.IMAGE_ROTATION_PREALIGN:
+            image_rotation_pre_align.run_image_rotation_pre_align(self._scanner, self._multibeam, self._descanner,
+                                                                  self._detector, self._dataflow, self._ccd,
+                                                                  self._det_rotator)
+            configure_asm(self._multibeam, self._descanner, self._detector, self._dataflow, self.asm_config,
+                          upload=False)
+
+        elif calibration == Calibrations.IMAGE_TRANSLATION_PREALIGN:
+            d_offset = image_translation_pre_align.run_image_translation_pre_align(self._scanner, self._multibeam,
+                                                                                   self._descanner, self._detector,
+                                                                                   self._dataflow, self._ccd)
+            self.asm_config["descanner"]["scanOffset"] = d_offset
+            configure_asm(self._multibeam, self._descanner, self._detector, self._dataflow, self.asm_config,
+                          upload=False)
+
+        elif calibration == Calibrations.IMAGE_ROTATION_FINAL:
+            image_rotation.run_image_rotation(self._scanner, self._multibeam, self._descanner,
+                                              self._detector, self._dataflow, self._det_rotator)
+            configure_asm(self._multibeam, self._descanner, self._detector, self._dataflow, self.asm_config,
+                          upload=False)
+
+        elif calibration == Calibrations.IMAGE_TRANSLATION_FINAL:
+            d_offset = image_translation.run_image_translation(self._scanner, self._multibeam, self._descanner,
+                                                               self._detector, self._dataflow, self._ccd)
+            self.asm_config["descanner"]["scanOffset"] = d_offset
+            configure_asm(self._multibeam, self._descanner, self._detector, self._dataflow, self.asm_config,
+                          upload=False)
+
+    def cancel(self, future):
+        """
+        Cancels the calibrations.
+        :param future: (future) The calibration future.
+        :return: (bool) True if cancelled, TODO False?
```

Comment:
I don't understand the TODO, can you add a bit more information?

Reply:
So far this function only ever returns True when called. I wonder if there is a case when it should return False. That would be the case when it is not possible to cancel right? But I cannot think about an event when it is not possible to cancel (maybe only at the very end?).

Reply:
I think it's always possible to cancel, indeed maybe at the very end that it just doesn't make sense anymore to actually cancel.

Reply:
Yeah, typically you don't need to worry too much. The Future.cancel() is supposed to return `False` if the future cannot be cancelled anymore. Typically it's either because it's is already finished (but that's handled by the parent function) or if it's so late that it's impossible to cancel the very last details. Usually, I don't bother implementing this. Typically, at the very end of the main run function, I check whether the Future was cancelled, and raise a CancelledError, which handles every case. In addition, even if you get this behaviour incorrect nothing will break, because nothing in Odemis relies on this.

Reply:
Makes sense and we check at the end of the run function if a CancelledError was raised. Thanks for explaining!